keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,258 @@
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Scalability and performance of kube-apiserver is crucial for scalability
+of the whole Kubernetes cluster. Given that kube-apiserver is cpu-intensive
+process, scaling a single instance of it translates to optimizing amount
+of work is needed to process a request (cpu cycles and amount of allocated
+memory, as memory management is significant part of work done be
+kube-apiserver).
+
+This proposal is aiming to significantly reduce amount of work spent on
+serializing objects as well as amount of allocated memory to process that.
+
+## Motivation
+
+Running different types of scalability tests and analyzing large production
+clusters proves that large number of watches watching the same set of objects
+may cause significant load on kube-apiserver. An extreme example of it is
+[#75294][], where creation of a single large Endpoints object (almost 1MB of
+size, due to 5k pods backing it) in 5k-node cluster can completely overload
+kube-apiserver for 5 seconds.
+
+The main reason for that is that for every watcher (Endpoints are being watched
+by kube-proxy running on every one) kube-apiserver independently serializes
+(which also requires deep-copy) every single object being send via this watch.
+
+While this problem is extremely visible for watch, the situation looks the same
+for regular GET/LIST operations - reading the same object N times will result
+in serializing that N times independently.
+
+This proposal presents a solution for that problem.
+
+[#75294]: https://github.com/kubernetes/kubernetes/issues/75294
+
+### Goals
+
+- Reduce load on kube-apiserver, by not serializing the same object multiple
```

Comment:
This makes a big assumption about our code, which is that our serialization is idempotent. In particular I can imagine some paths getting e.g. a self link wrong or different. What can we do to ensure that all serialization entrypoints produce the exact same serialized bytes?

Reply:
> In particular I can imagine some paths getting e.g. a self link wrong or different

Not to derail your general point, but do we think anyone has actually needed and used the self-link.  They were broken for many release in a row a while back....

Reply:
Good question, I'm open to removing it, but it would be a significant undertaking...

Reply:
> This makes a big assumption about our code, which is that our serialization is idempotent.

I'm not sure I understand this question - object once serialized, can't be serialized again (it's no longer object, but []byte). Can you clarify?

Regarding selflinks - they are quite painful indeed (it's the only think we set further in the stack), so if we could get rid of those, it would make things easier.

Reply:
No, I'm concerned that objects egressing apiserver via different routes might have been serialized differently, and that would stop being true. It sounds like a bug when I phrase it like that...

Reply:
Yeah - I think I'm still confused - if we can get different thing it sounds like a bug to me...

Reply:
self links *are* different for different endpoints that serve the same group/version/kind:

```
$ kubectl get --raw /api/v1/namespaces/kube-system/pods/kube-dns-5994d786b-rtnpq | jq .metadata.selfLink
"/api/v1/namespaces/kube-system/pods/kube-dns-5994d786b-rtnpq"

$ kubectl get --raw /api/v1/namespaces/kube-system/pods/kube-dns-5994d786b-rtnpq/status | jq .metadata.selfLink
"/api/v1/namespaces/kube-system/pods/kube-dns-5994d786b-rtnpq/status"
```

Reply:
Ah, of course, thank you, let's pretend that this is what I had in mind rather than just a poorly articulated suspicion :)

Reply:
Thanks Jordan - I didn't know about that.
Are those different depending on subresource only, or they can differ also in other situations?

Reply:
subresource is the only one that comes to mind, but it's more a matter of the information only being available at output time (in [transformObject](https://github.com/kubernetes/kubernetes/blob/c28b3b1fdd0d00972461653ac33599ffe5eccb2e/staging/src/k8s.io/apiserver/pkg/endpoints/handlers/response.go#L40-L46))

Reply:
I don't think this is problem for my proposal - where serializing is done at the very end (when self-links are already set) [so I'm proposing to do serialization (or get data from cache] when regular serialization is done]. 

What I just realized is that actually, given that SelfLink reflects both group0-version and subresource, maybe we can actually do caching based on SelfLink, which would solve this particular issue (caching is done when all the information is already set).

Reply:
Using selfLink as the key for different serializations of a given object sounds clever.

Reply:
The more I think about it, the more I'm convinced that something like that will be needed in every single approach I can imagine.
I will try to update the proposal in the next couple days.

Reply:
I updated the proposal to reflect the "SelfLink"-related discussion.